internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,255 @@
+// One logGatherer will run per log target. Its loop() method should be run
+// in its own goroutine. It can be stopped in a separate goroutine by calling
+// the stop() method.
+// The logGatherer will "flush" the client:
+//   - on a regular cadence (e.g. every 1 second)
+//   - when it is told to shut down.
+//
+// Its client may also flush itself when its internal buffer reaches a certain
+// size.
+type logGatherer struct {
+	targetName string
+	tickPeriod time.Duration
+	client     logClient
+
+	entryCh     chan servicelog.Entry
+	pullers     map[string]*logPuller
+	pullersLock sync.Mutex
+
+	// Context to pass to client
+	// When we are asked to shutdown, we can cancel this context to ensure the
+	// client is not blocking us.
+	clientCtx    context.Context
+	clientCancel context.CancelFunc
+
+	// tomb for the main loop
+	tomb *tomb.Tomb
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	g := &logGatherer{
+		targetName: target.Name,
+		tickPeriod: tickPeriod,
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+		tomb:       &tomb.Tomb{},
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	return g, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("plan changed")
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	if puller, ok := g.pullers[service.Name]; ok {
+		// This should never happen, but just in case, shut down the old puller.
+		_ = puller.tomb.Killf("service started")
+	}
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
+}
+
+func (g *logGatherer) loop() error {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-g.tomb.Dying():
+			return nil
+
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush(g.clientCtx)
```

Comment:
Could these calls use the tomb's context, as in `g.client.Flush(g.tomb.Context(nil))`? I wonder if we'd need `clientCtx` and `clientCancel` at all in that case -- when `stop` kills the tomb, I believe the context will be cancelled.

Reply:
Hmm, I think the issue is:
- when we call `gatherer.stop`, we want it to stop pulling logs right away
- But we want to allow a grace period (e.g. 5 seconds) to flush out the existing logs

So I think we need to distinguish between:
- a soft stop signal (i.e. "hey it's time to start shutting down") - this is the `tomb.Kill`
- a hard stop signal (i.e. "alright I'm pulling the plug, stop right now") - this is the `clientCancel`

As far as I know, `tomb` doesn't really give you a way to model this. We can discuss whether this is the correct behaviour though.

Reply:
I've rewritten the `gatherer.stop` method to handle edge cases, and added comments throughout. Hopefully this makes it clear why (I think) we need separate contexts here.